Add a document outlining how to build native packages locally. (#12431)
netdata/netdata: packaging/building-native-packages-locally.md

## Requirements

You will need:

* A working Docker or Podman host.
* A local copy of the source tree you want to build from.

## Building the packages

In the root of the source tree you want to build from, clean up any existing files left over from a previous build
and then run:

```bash
docker run -it --rm -e VERSION=0.1 -v $PWD:/netdata netdata/package-builders:<tag>
```

or

```bash
podman run -it --rm -e VERSION=0.1 -v $PWD:/netdata netdata/package-builders:<tag>
```

The `<tag>` should be the lowercase distribution name with no spaces, followed by the
release of that distribution. For example, `centos7` to build on CentOS 7, or `ubuntu20.04`
to build on Ubuntu 20.04. Note that we use Alma Linux for builds on CentOS/RHEL 8 or newer. See
[netdata/package-builders](https://hub.docker.com/r/netdata/package-builders/tags) for all available tags.

Once it finishes, the built packages can be found under `artifacts/` in the source tree.

If an error is encountered and the build is being run interactively, it will drop to a shell to allow you to
inspect the state of the container and look at build logs.

### Detailed explanation

> The environments used for building our packages are fully self-contianed Docker images built from the Dockerfiles

Reviewer: ```suggestion
The environments used for building our packages are fully self-contianed Docker images built from [Dockerfiles](https://github.com/netdata/helper-images/tree/master/package-builders)
```

> located at https://github.com/netdata/helper-images/tree/master/package-builders. These are published on Docker

Reviewer: ```suggestion
These are published on Docker
```

Hub with the image name `netdata/package-builders`, and tagged using the name and version of the distribution
(with the tag corresponding to the suffix on the associated Dockerfile).

> The build code has two specific requirements when run:

Reviewer: ```suggestion
The build code has two specific requirements when run:
```
Can we say "The build code expects the following requirements to bet met:"

- It expects the source tree it should build from to be located at `/netdata`, and expects that said source tree
  is clean (no artifacts left over from previous builds).
- It expects an environment variable named `VERSION` to be defined, and uses this to control what version number
  will be shown in the package metadata and filenames.

Internally, the source tree gets copied to a temporary location for the build process so that the souce tree can
be mounted directly from the host without worrying about leaving a dirty tree behind, any templating or file
movements required for the build to work are done, the pakcage build command is invoked with the correct arguments,

> and then the resultant packages are copied to the `artifacts/` directory in the original sourc etree so they are

Reviewer: ```suggestion
and then the resultant packages are copied to the `artifacts/` directory in the original source tree so they are
```

<!--
title: How to build native (DEB/RPM) packages locally for testing
description: Instructions for developers who need to build native packages locally for testing.
custom_edit_url: https://github.com/netdata/netdata/edit/master/packaging/building-native-packages-locally.md
keywords: [Netdata native package, Netdata RPM, Netdata DEB, Testing native packages Netdata]
-->

# How to build native (DEB/RPM) packages locally for testing

## Requirements

> You will need:

Reviewer: ```suggestion
To build native packages locally, you will need the following:
```

* A working Docker or Podman host.
* A local copy of the source tree you want to build from.

## Building the packages

> In the root of the source tree you want to build from, clean up any existing files left over from a previous build

Reviewer: ```suggestion
In the root of the source tree from which you want to build, clean up any existing files left over from a previous build
```

being built. As a general rule, it needs to start with a digit, and must include a `.` somewhere.

Once it finishes, the built packages can be found under `artifacts/` in the source tree.

If an error is encountered and the build is being run interactively, it will drop to a shell to allow you to
inspect the state of the container and look at build logs.

### Detailed explanation

The environments used for building our packages are fully self-contianed Docker images built from [Dockerfiles](https://github.com/netdata/helper-images/tree/master/package-builders)
These are published on Docker
Hub with the image name `netdata/package-builders`, and tagged using the name and version of the distribution
(with the tag corresponding to the suffix on the associated Dockerfile).

The build code expects the following requirements to be met:

- It expects the source tree it should build from to be located at `/netdata`, and expects that said source tree
  is clean (no artifacts left over from previous builds).
- It expects an environment variable named `VERSION` to be defined, and uses this to control what version number
  will be shown in the package metadata and filenames.

> Internally, the source tree gets copied to a temporary location for the build process so that the souce tree can

Reviewer: ```suggestion
Internally, the source tree gets copied to a temporary location for the build process so that the source tree can
```

be mounted directly from the host without worrying about leaving a dirty tree behind, any templating or file

> movements required for the build to work are done, the pakcage build command is invoked with the correct arguments,

Reviewer: ```suggestion
movements required for the build to work are done, the package build command is invoked with the correct arguments,
```

and then the resultant packages are copied to the `artifacts/` directory in the original source tree so they are
accessible after the container exits.

## Finding build logs after a failed build

Build logs and artifacts can be found in the build directory, whose location varies by distribution.

On DEB systems (Ubuntu and Debian), the build directory inside the container is located at `/usr/src/netdata`

On RPM systems except openSUSE, the build directory inside the container is located under `/root/rpmbuild/BUILD/`
and varies based on the package version number.

On openSUSE, the build directory inside the container is located under `/usr/src/packages/BUILD`and varies based
on the package version number.

## Building for other architectures

If you need to test a build for an architecture that does not match your host system, you can do so by setting up

> QEMU user-mode emulation. This requires a Linux kernel with binfmt\_misc support (all modern distributions provide

Reviewer: I think the backslash is a typo. Also, can we research and remove the parenthetical here? It reads like an editing note.

Author: No, the backslash is intention here, it’s explicitly escaping the `_` so it doesn’t mistakenly get matched elsewhere in the file and result in text unintentionally becoming italicized (note that in the rendered version, it’s not visible).

Regarding the parenthetical, we probably should check WSL and Docker Desktop, but I think it makes more sense to just drop those from the comment altogether.

Reviewer: Agree on dropping them.

this out of the box, but I’m not sure about WSL or Docker Desktop).

> The quick and easy way to do this is to run:

Reviewer: ```suggestion
The quick and easy way to do this is to run the following:
```

```bash
docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
```

or

```bash
podman run --rm --privileged multiarch/qemu-user-static --reset -p yes
```

> Which will set up the required QEMU user-mode emulation until you reboot. Note that if using Podman, you will need

Reviewer: ```suggestion
This will set up the required QEMU user-mode emulation until you reboot. Note that if using Podman, you will need
```

> # How to build native (DEB/RPM) packages locally for testing

Reviewer 2: ```suggestion
<!--
title: How to build native (DEB/RPM) packages locally for testing
description: 
custom_edit_url: https://github.com/netdata/netdata/edit/master/packaging/building-native-packages-locally.md
keywords: [Netdata native package, Netdata RPM, Netdata DEB, Testing native packages Netdata]
-->

# How to build native (DEB/RPM) packages locally for testing

> Please provide a short summary
```
@Ferroin 
Please provide a fitting `description` and a short summary. It can be the same, since `description` is the snippet that search engines preview - so it's not necessarily visible for the user.

> 

Reviewer 2: ```suggestion
## Requirements
```

<!--
title: How to build native (DEB/RPM) packages locally for testing
description: 
custom_edit_url: https://github.com/netdata/netdata/edit/master/packaging/building-native-packages-locally.md
keywords: [Netdata native package, Netdata RPM, Netdata DEB, Testing native packages Netdata]
-->

# How to build native (DEB/RPM) packages locally for testing

> > Please provide a short summary

Reviewer 2: @Ferroin
Please provide a fitting description and a short summary. It can be the same, since description is the snippet that search engines preview - so it's not necessarily visible for the user.

<!--
title: How to build native (DEB/RPM) packages locally for testing
description: 
custom_edit_url: https://github.com/netdata/netdata/edit/master/packaging/building-native-packages-locally.md
keywords: [Netdata native package, Netdata RPM, Netdata DEB, Testing native packages Netdata]
-->

# How to build native (DEB/RPM) packages locally for testing

> Please provide a short summary

## Building the native Netdata package 

### Requirements

You will need:

* A working Docker host (it _may_ work with Podman or Singularity, but they are untested).
* A local copy of the source tree you want to build from.

To build a native package from source:

1. Prepare the directory: Clean up any existing files left over from a previous build in the root of the source tree you want to build from. 

> 2. To build the package run the command below. The `<tag>` should be the lowercase distribution name with no spaces, followed by the release of that distribution. For example, `ubuntu20.04` to build on Ubuntu 20.04, or  `centos7` to build on CentOS 7 (use `rockylinux8` for  CentOS 8 or RHEL 8).

Reviewer 3: ```suggestion
2. To build the package run the command below. The `<tag>` should be the lowercase distribution name with no spaces, followed by the release of that distribution. For example, `ubuntu20.04` to build on Ubuntu 20.04, or  `centos7` to build on CentOS 7 (use `rockylinux8` for  CentOS 8 or RHEL 8). See [netdata/package-builders](https://hub.docker.com/r/netdata/package-builders/tags) for all available tags.
```

The build code has two specific requirements when run:

* It expects the source tree it should build from to be located at `/netdata`, and expects that said source tree
  is clean (no artifacts left over from previous builds).
* It expects an environment variable named `VERSION` to be defined, and uses this to control what version number
  will be shown in the package metadata and filenames.

When you run the package builder the following happens: 
1. **Preparation**: The source tree gets copied to a temporary location for the build process. This allows the system to mount the souce tree directly from the host without worrying about leaving a dirty tree behind. Then the script creates any templates or moves files as needed to build the package.
2. **Building**: The package build command is invoked with the correct arguments and the package is built.
3. **Permanent storage**: The resulting packages are copied to the `artifacts/` directory in the original source tree so they are

> accessible after the container exits.

Author: I really don’t know that we need the ‘step-by-step’ organization here. It feels out of place IMO given that we are not instructing the user to do anything

Reviewer 2: I disagree to an extent. We are explaining a workflow that is processed in a certain order, so making these different phases more visible helps IMO. 

However, we can debate if it's really necessary to use an ordered list. 
We could go for unordered or plain paragraphs if you prefer.

Author: Hmm, I think the issue here that makes it feel so out of place to me is that it feels like prosaic creative writing the way it’s presented here, not technical documentation. IOW, the structure makes it sound to me less like we’re trying to unambiguously explain a process, and more like we’re trying to make the process sound ‘cool’, and possibly like we’re trying to cater to end users (who are not the target audience for this documentation).

Reviewer 2: @DShreve2 What's your take here?

* A local copy of the source tree you want to build from.

To build a native package from source:

1. Prepare the directory: Clean up any existing files left over from a previous build in the root of the source tree you want to build from. 
2. To build the package run the command below. The `<tag>` should be the lowercase distribution name with no spaces, followed by the release of that distribution. For example, `ubuntu20.04` to build on Ubuntu 20.04, or  `centos7` to build on CentOS 7 (use `rockylinux8` for  CentOS 8 or RHEL 8).

   ```bash
   docker run -it --rm -e VERSION=0.1 -v $PWD:/netdata netdata/package-builders:<tag>
   ```
   🟢 Once it finishes, the built packages can be found under `artifacts/` in the source tree.
   
   🔴 If you encounter an error: Run the build interactively (`--interactive`) to inspect the state of the container and look at build logs.

## Building for other architectures

### Requirements

If you want to test a build for an architecture that does not match your host system, you can do so by setting up
**QEMU user-mode emulation**. This requires a Linux kernel with binfmt_misc support which all modern distributions provide

> this out of the box (support for WSL and Docker Desktop unknown).

Author: ```suggestion
support out of the box (support for WSL and Docker Desktop unknown).
```
